vms/platformvm/state/stakers.go

```diff
@@ -135,98 +166,228 @@ func (v *baseStakers) PutValidator(staker *Staker) {
+	validator.validator = staker
+
+	// Explicitly remove prevStaker from stakers tree. This is because stakers tree
+	// identifies stakers via stakers.Less function, so a staker with updated Start/EndTime
+	// would be treated as a different staker and not be replaced by ReplaceOrInsert call.
+	v.stakers.Delete(prevStaker)
 	v.stakers.ReplaceOrInsert(staker)
+
+	validatorDiff := getOrCreateDiff(v.validatorDiffs, staker.SubnetID, staker.NodeID)
+	if validatorDiff.validator.status == added {
+		// validator was added and is being immediately updated.
+		// This should not currently happen, but for sake of generality,
+		// we mark it as added as we do for diffs.
+		validatorDiff.validator = stakerAndStatus{
+			staker: staker,
+			status: added,
+		}
+	} else {
+		validatorDiff.validator = stakerAndStatus{
+			staker: staker,
+			status: updated,
+		}
+	}
+
+	return nil
 }
 
 func (v *baseStakers) DeleteValidator(staker *Staker) {
-	validator := v.getOrCreateValidator(staker.SubnetID, staker.NodeID)
+	var (
+		subnetID = staker.SubnetID
+		nodeID   = staker.NodeID
+	)
+	validator, found := v.getValidator(subnetID, nodeID)
+	if !found {
+		// deleting an non-existing staker. Nothing to do.
+		return
+	}
+	storedStaker := validator.validator
 	validator.validator = nil
-	v.pruneValidator(staker.SubnetID, staker.NodeID)
-
-	validatorDiff := v.getOrCreateValidatorDiff(staker.SubnetID, staker.NodeID)
-	if validatorDiff.validatorStatus == added {
-		validatorDiff.validatorStatus = unmodified
-		validatorDiff.validator = nil
+	v.pruneValidator(subnetID, nodeID)
+
+	// for sake of generality, we assume we could insert and immediately delete
+	// a validator.validator. We need to handle this case separately to avoid
+	// messing validators set.
+	// Also, for sake of generality, we assume we could delete an updated version
+	// of validator.validator. We explicitly remove the currently stored
+	// version of the staker to handle this case.
+	validatorDiff := getOrCreateDiff(v.validatorDiffs, subnetID, nodeID)
+	if validatorDiff.validator.status == added {
+		validatorDiff.validator = stakerAndStatus{
+			staker: nil,
+			status: unmodified,
+		}
 	} else {
-		validatorDiff.validatorStatus = deleted
-		validatorDiff.validator = staker
+		validatorDiff.validator = stakerAndStatus{
+			staker: storedStaker,
+			status: deleted,
+		}
 	}
 
-	v.stakers.Delete(staker)
+	v.stakers.Delete(storedStaker)
 }
 
 func (v *baseStakers) GetDelegatorIterator(subnetID ids.ID, nodeID ids.NodeID) StakerIterator {
-	subnetValidators, ok := v.validators[subnetID]
-	if !ok {
-		return EmptyIterator
-	}
-	validator, ok := subnetValidators[nodeID]
-	if !ok {
+	validator, found := v.getValidator(subnetID, nodeID)
+	if !found {
 		return EmptyIterator
 	}
-	return NewTreeIterator(validator.delegators)
+	return NewTreeIterator(validator.sortedDelegators)
 }
 
 func (v *baseStakers) PutDelegator(staker *Staker) {
+	// Note: a delegator may be inserted before its validator
```

Comment:
this is due to delegator/validator priority if AddValTx/AddDelTx are included in the very same block